### ISSUE TITLE
presign object retrieval 500 errors - is retry logic needed?

### ISSUE THREAD

**Issue by terrywbrady** [CDLUC3/mrt-doc#1611]
Example error from Access01.

Note that our tomcat json logs do not capture details.  This is from the tomcat access log.

```
172.30.26.243 - - [22/Sep/2023:10:12:11 -0700] "GET /presign-obj-by-token/ad0d87aa-4474-4bea-b132-d1ce1ca23782?contentDisposition=attachment%3B+filename%3Ddoi_10.5061_dryad.75r70__v1.zip HTTP/1.1" 500 514
```

**Comment by dloy** [comment]
# mrt-cloud problem

## Problem
Right now the storage server is issuing periodic 500 status responses when trying to access the S3 token object.

__Note:__ the term object here corresponds to an S3 saved file.

After doing some research using the catalina.out against localhost_access_log and storage().log it became clear that the 500 status codes in the access log corresponded to GENERAL_EXCEPTION errors in catalina.out. Fortunately the catalina error has a full command trace. The command trace indicates the error is caused by the S3 TransferManager failing to find the content to download.

The S3 Get process has 2 steps:
- metadata request to see if content exists
- download content if exists

The S3 Put process has the following steps:
- does object already exist (metadata request)
- match saved object digest against new object (file) to be added
- if same object return
- if different object then delete object
- add new object

The delete in the S3 Put is used to prevent multiple versions of a file if multi-versioning is allowed on a bucket.

In an asynchronous environment where the same object is frequently replaced at different steps (token zip handling)  with multiple threads, the error occurs  when an S3 Put deletes an object between S3 get metadata (which works) and the get read (which fails).

## Solution
Remove the delete from the S3 Put. If the S3 bucket is multi-version then an extra object will exist but the process works. If the S3 bucket is single version then the object is replaced. In either case the object will exist for download and the S3 Get will succeed (no 500).

Additional changes to the mrt-cloud code will share a common timeout and test logic regardless of the S3 service (minio, wasabi, aws). Right now the s3client is handled differently for each service type.

View this as fairly high priority.

**Comment by dloy** [comment]
The 500 errors occur because of asynch issues between submission of the initialization of the token file and the attempted read of the token file by the secondary zookeeper process to start building the zip file. An error in the cloud storage failed to capture the 404. The 500 error will be replaced with the accurate 404. The processing continues to completion. 200

**Comment by elopatin-uc3** [comment]
Thanks for the investigation and explanation @dloy, and for the fix in the upcoming release (1.5 tag).
